---
title: Prioiritized list of fine mapping data according Dr.William Wen's paper
author: Charles Zhou
date: '2022-04-22'
slug: analysis
categories:
  - analysis
tags: []
output: md_document
---

# Read in the data from GETx project ([data is from Dr.William Wen's paper](https://gtexportal.org/home/datasets)).

```{r, include=TRUE,echo=TRUE,message=FALSE,warning=FALSE}

library(tidyverse)
library(dplyr)
library(data.table)

setwd("D:/Charles_coursework/PME/Rotation/Haky_lab/singleXcan/singleXcan_project/GTEx_v8_finemapping_DAPG/")
getwd()


original <- read_table("GTEx_v8_finemapping_DAPG.txt")

tissue <- unique(original$tissue_id)
length(tissue)

vcf3 <- fread("GTEx_v8_finemapping_DAPG.vcf")
CS95 <- read_table("GTEx_v8_finemapping_DAPG.CS95.txt")

head(CS95)
head(original)
head(vcf3)

```

# Get the prioritized list of the variants
Filter the variant by the cluster pip (posterior inclusion probability) and variant pip value
```{r,include=TRUE,echo=TRUE,message=FALSE,warning=FALSE}


prioritized <- original %>%
  filter(cluster_pip >= 0.8) %>%
  group_by(tissue_id, cluster_id, gene_id) %>%
  arrange(desc(variant_pip), .by_group = TRUE) %>%
  mutate(rank = dense_rank(desc(variant_pip))) %>%
  filter(rank == 1)

# write.csv(prioritized, file = "prioritized_list.csv")

head(prioritized)
```
# Take a look at some plots drawn from the data
## Variant pip distribution of different tissues within prioiritized data
```{r,include=TRUE,echo=TRUE,message=FALSE,warning=FALSE}
pri <- prioritized %>% filter(tissue_id %in% c("Lung","Thyroid", "Testis", "Whole_Blood", "Pancreas")) %>% ungroup()

pri %>% ggplot(aes(variant_pip)) +
  geom_histogram() + facet_wrap(~tissue_id)
  
pri %>% count(gene_id, variant_id) %>% arrange(desc(n))
```
## Information about variant_pip value distribution within a tissue (pancreas as an example)
Take a look at the sum of the variant_pip value of each gene-cluster pair in pancreas as an example

```{r,include=TRUE,echo=TRUE,message=FALSE,warning=FALSE}
pancreas <- original %>%
  filter(tissue_id == "Pancreas")
  
pancreas_modi <- pancreas %>%
  group_by(cluster_id, gene_id) %>%
  # group_by(cluster_id) %>%
  # arrange(desc(variant_pip), .by_group = TRUE) %>%
  summarise(sum(variant_pip), .groups = "rowwise") %>%
  arrange(desc(`sum(variant_pip)`), .by_group = TRUE)

pancreas_modi <- pancreas_modi %>%
  mutate(rank = dense_rank(desc(`sum(variant_pip)`))/length(pancreas_modi$gene_id)) %>%
  ungroup()

pancreas_modi <- pancreas_modi %>%
  mutate(rank = dense_rank(desc(`sum(variant_pip)`))/length(pancreas_modi$gene_id)) %>%
  ungroup()


ggplot(data = pancreas_modi) +
  geom_point(mapping = aes(x = rank , y = `sum(variant_pip)`) ) +
  xlab("proportion")+
  ylab("sum of the variant_pip value of each gene-cluster pair")+
  ggtitle("pancreas")
```

Take pancreas tissue as an example to look at the variant_pit distribution:

```{r,echo=TRUE,message=FALSE,warning=FALSE}
pancreas <- pancreas %>%
  # group_by(cluster_id, gene_id) %>%
  # group_by(cluster_id) %>%
  # arrange(desc(variant_pip), .by_group = TRUE) %>%
  mutate(proportion = dense_rank(desc(variant_pip))/length(pancreas$gene_id)) %>%
  ungroup()
  


ggplot(data = pancreas) +
  geom_point(mapping = aes(x = proportion , y = variant_pip)) 

ggplot(data = pancreas) +
  geom_histogram(mapping = aes(x = variant_pip, y = stat(count/sum(count))))

```

## Max variant_pip value of different tissues

The top 5000 variant_pip values of each tissue are averaged (if the number of the samples chosen from each sample is too small, the average variant_pip should be all closed to 1). Then the tissues are sorted by the descent order of average top 5000 variant_pip value.

```{r,echo=TRUE,message=FALSE,warning=FALSE}
tissue_sample <- original %>%
  group_by(tissue_id) %>%
  arrange(desc(variant_pip), .by_group = TRUE) %>%
  top_n(5000) %>%
  summarise(mean(variant_pip)) %>%
  arrange(desc(`mean(variant_pip)`), .by_group = TRUE)


tissue_plot <- ggplot(data = tissue_sample) +
  geom_col(mapping = aes(x = reorder(tissue_id, + `mean(variant_pip)`), y = `mean(variant_pip)`))+
  theme(axis.text.x = element_text(angle = 90, vjust = 0, hjust= 0, color = 'Black'))+
  coord_flip()+
  ggtitle("Average of top 5000 variant_pip of different tissues")

tissue_plot
```

